### PR TITLE
feat(plugins): ai-proxy-plugin

### PR DESCRIPTION
**Pull request by tysoekong**
### Summary

This commit transforms and proxies requests to a number of AI / LLM providers and models.

It adds a plugin that takes requests in one of a few defined and standardised formats, translates them to the configured target format, and then transforms the response back into a standard format.

The request/response formats are based on OpenAI, and [conform to the JSON schema in the fixtures](https://github.com/Kong/kong/blob/cef856da6b6ad3bad18d74a659003fc489a2a997/spec/03-plugins/38-ai-proxy/ai-proxy/json-schema.json)

[There is a sample API specification to describe the suported formats also](https://github.com/ttyS0e/kong/blob/master/spec/fixtures/ai-proxy/oas.yaml)

The plugin controls:

1. The backend URL (unless it's a self-hosted model, in which case set `config.model.options.upstream_url`)
2. The API key insertion
3. The request/response body transformation
4. Capturing and storing metrics from the upstream responses, into normalised `kong.log` entries, which will then output via any configured logging platform e.g. file-log, kafka-log...

The reasoning for trying to flatten all provider formats, is that it allows standardised manipulation of the data before and after transmission.

It currently supports `v1/chat` and `v1/completion` style requests for all available providers. Each provider implementation class is in its own module under `kong.plugins.ai-proxy.drivers`.

You can also set mode `preserve` to pass through, and the plugin will perform a best-effort to conform to the chosen `config.model.provider`, based on the configured header patterns and values. *This may need re-thinking, please provide input on this.*

This implementation only supports REST-based full text responses, but streaming support could be added based upon adoption, as with Kong >3.0 we can now process WebSocket frames.

Ideal documentation (docs.konghq.com) is being finalised internally, will be ready soon.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### CODE REVIEW

**Review by flrgh — APPROVED** [review]

**Comment by tysoekong** [comment]
@fffonion Anything else for me to do here?

File: kong/llm/init.lua [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+local fmt      = string.format
+local cjson    = require("cjson.safe")
+local re_match = ngx.re.match
+
+local ai_shared = require("kong.llm.drivers.shared")
+--
+
+local _M = {}
+
+local auth_schema = {
```
Comment (locao) [inline kong/llm/init.lua:12]: The schemas could go to their own file to keep this cleaner.

File: kong/llm/init.lua [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+  if not new_request_body then
+    return nil, "no response choices received from upstream AI service"
+  end
+
+  -- if specified, extract the first regex match from the AI response
+  -- this is useful for AI models that pad with assistant text, even when
+  -- we ask them NOT to.
+  if response_regex_match then
+    local matches, err = re_match(new_request_body, response_regex_match, "ijm")
+    if err then return nil, "failed regex matching ai response: " .. err end
```
Comment (locao) [inline kong/llm/init.lua:314]: Nitpick: formatting
```suggestion
    if err then
      return nil, "failed regex matching ai response: " .. err
    end
```

File: kong/llm/init.lua [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+    end
+  end
+
+  return new_request_body
+end
+
+function _M:parse_json_instructions(body_string)
+  local instructions, err = cjson.decode(body_string)
+  if err then return nil, nil, nil, err end
```
Comment (locao) [inline kong/llm/init.lua:330]: Nitpick: formatting
```suggestion
  if err then
    return nil, nil, nil, err
  end
```

**Review by locao — APPROVED** [review]
This is great! 🚀 

None of my comments are real blockers to merge this PR, just some suggestions.

**Pull request merged**